test/Microsoft.Extensions.Internal.Test/ObjectMethodExecutorTest.cs

```diff
@@ -0,0 +1,390 @@
+            Assert.True(executor.IsMethodAsync);
+            Assert.Same(typeof(TestObject), executor.AsyncResultType);
+            Assert.NotNull(result);
+            Assert.Equal("Hello 123", result.value);
+        }
+
+        [Fact]
+        public async void TargetMethodReturningCustomAwaitableOfValueType_CanInvokeViaExecute()
+        {
+            // Arrange
+            var executor = GetExecutorForMethod("CustomAwaitableOfValueTypeAsync");
+
+            // Act
+            var result = await (TestAwaitable<int>)executor.Execute(_targetObject, new object[] { 123, 456 });
+
+            // Assert
+            Assert.True(executor.IsMethodAsync);
+            Assert.Same(typeof(int), executor.AsyncResultType);
+            Assert.NotNull(result);
+            Assert.Equal(579, result);
+        }
+
+        [Fact]
+        public async void TargetMethodReturningCustomAwaitableOfReferenceType_CanInvokeViaExecuteAsync()
+        {
+            // Arrange
+            var executor = GetExecutorForMethod("CustomAwaitableOfReferenceTypeAsync");
+
+            // Act
+            var result = await executor.ExecuteAsync(_targetObject, new object[] { "Hello", 123 });
+
+            // Assert
+            Assert.True(executor.IsMethodAsync);
+            Assert.Same(typeof(TestObject), executor.AsyncResultType);
+            Assert.NotNull(result);
+            Assert.IsType(typeof(TestObject), result);
+            Assert.Equal("Hello 123", ((TestObject)result).value);
+        }
+
+        [Fact]
+        public async void TargetMethodReturningCustomAwaitableOfValueType_CanInvokeViaExecuteAsync()
+        {
+            // Arrange
+            var executor = GetExecutorForMethod("CustomAwaitableOfValueTypeAsync");
+
+            // Act
+            var result = await executor.ExecuteAsync(_targetObject, new object[] { 123, 456 });
+
+            // Assert
+            Assert.True(executor.IsMethodAsync);
+            Assert.Same(typeof(int), executor.AsyncResultType);
+            Assert.NotNull(result);
+            Assert.IsType(typeof(int), result);
+            Assert.Equal(579, (int)result);
+        }
+
+        [Fact]
+        public async void TargetMethodReturningAwaitableOfVoidType_CanInvokeViaExecuteAsync()
+        {
+            // Arrange
+            var executor = GetExecutorForMethod("VoidValueMethodAsync");
+
+            // Act
+            var result = await executor.ExecuteAsync(_targetObject, new object[] { 123 });
+
+            // Assert
+            Assert.True(executor.IsMethodAsync);
+            Assert.Same(typeof(void), executor.AsyncResultType);
+            Assert.Null(result);
+        }
+
+        private ObjectMethodExecutor GetExecutorForMethod(string methodName)
+        {
+            var method = typeof(TestObject).GetMethod(methodName);
+            var executor = ObjectMethodExecutor.Create(method, targetTypeInfo);
+            return executor;
+        }
+
+        public class TestObject
```

Comment:
Could you throw in a test using `ValueTask`? It should be covered by the custom awaitables, but that's one of the primary new types we want to support :).